### PR TITLE
validate calls to DrawingContext

### PR DESCRIPTION
**Pull request by danwalmsley**
## What does the pull request do?
If you get invalid numbers like double.NaN or double.Infinity, we completely break the render pass, and its very difficult for the developer to work out where the invalid call came from.

This add validation and logging.

@jmacato this probably need to be added for the transitions too? remember we have invalid numbers crop up there too? Not sure where that is though.

## What is the current behavior?

below shows left, avalonia and right wpf. When rendering a button and random lines, but also some of the drawline calls have points with `double.NegativeInfinity`, it completely destroys the rendering of the button.
render-failures

## What is the updated/expected behavior with this PR?
Dont fail, and continue to render other controls, like wpf, also log warnings so the developer would be able to know about it.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

### CODE REVIEW

File: src/Avalonia.Visuals/Media/DrawingContext.cs [hunk 1]
```diff
@@ -8,6 +9,29 @@
...
+        {
+            return !(double.IsNaN(d) || double.IsInfinity(d));
+        }
+
+        public static bool IsRenderValid (this Point p)
+        {
+            return p.X.IsRenderValid() && p.Y.IsRenderValid();
+        }
+
+        public static bool IsThicknessValid (this Thickness t)
```
Comment (kekekeks) [inline src/Avalonia.Visuals/Media/DrawingContext.cs:24]: inconsistent naming

File: src/Avalonia.Visuals/Media/DrawingContext.cs [hunk 2]
```diff
@@ -109,7 +133,14 @@ public void DrawLine(IPen pen, Point p1, Point p2)
...
             if (PenIsVisible(pen))
             {
-                PlatformImpl.DrawLine(pen, p1, p2);
+                if (pen.IsRenderValid() && p1.IsRenderValid() && p2.IsRenderValid())
+                {
+                    PlatformImpl.DrawLine(pen, p1, p2);
+                }
+                else
+                {
+                    Logger.TryGet(LogEventLevel.Warning)?.Log("DrawLine", this, "Invalid Draw Parameters");
```
Comment (kekekeks) [inline src/Avalonia.Visuals/Media/DrawingContext.cs:142]: Why "DrawLine"? It's supposed to be an area

Comment (danwalmsley) [reply src/Avalonia.Visuals/Media/DrawingContext.cs]: fixed

**Comment by Gillibald** [comment]
I don't like this kind of validation. We need a way to have these checks only with debug builds and don't magicly set values to 0 and instead throw an exception. 

These invalid values should never be produced. Let's find out where they come from. 

For Thickness etc. it makes sense to validate on construction.

**Comment by grokys** [comment]
I too don't understand the root cause here. The comment on the wasabi issue says:

> it is actually possible for drawing contexts to be passed nan or infinity and it will completely blank your screen

Which calls in `DrawingContext` will do this? Trying to e.g. draw a rectangle with `NaN`?

**Comment by grokys** [comment]
> These invalid values should never be produced. Let's find out where they come from.

If the scene is completely messed up by calling e.g. `context.FillRectangle(Brushes.Red, new Rect(double.NaN, 0, 100, 100))` then I don't think we can stop these values being produced.

But I find it strange that this corrupts the whole scene. What is happening to get the whole scene corrupted?

**Comment by danwalmsley** [comment]
> > These invalid values should never be produced. Let's find out where they come from.
> 
> If the scene is completely messed up by calling e.g. `context.FillRectangle(Brushes.Red, new Rect(double.NaN, 0, 100, 100))` then I don't think we can stop these values being produced.
> 
> But I find it strange that this corrupts the whole scene. What is happening to get the whole scene corrupted?

I did a test, and with Direct2D backend it doesnt happen.

On Skia it does, so I assume Direct2D just says, NaN or Infinity is impossible to draw, and so produces nothing, but silently skips.

iv added a repro here: https://github.com/AvaloniaUI/Avalonia/tree/repro-invalid-draw-calls

Iv seen quite a few times this symptom happen and no one can ever explain it or repro it, and you have zero chance of debugging it.

**Comment by danwalmsley** [comment]
> I don't like this kind of validation. We need a way to have these checks only with debug builds and don't magicly set values to 0 and instead throw an exception.
> 
> These invalid values should never be produced. Let's find out where they come from.
> 
> For Thickness etc. it makes sense to validate on construction.

If it was cut out of debug builds end users couldnt get the logger warnings... however we could only do the validation if a logger is attached? then it would be zero cost unless a logger was installed?

**Comment by danwalmsley** [comment]
These invalid values should never be produced. Let's find out where they come from.

we probably have some internal controls that can in extreme states do this, forexample iv seen itemsrepeater make the screen flicker blank.

 other than that users can implement controls that do this, and their UI will just start acting up... they will just say, oh Avalonia is not working,

If we can have a cost free validation with logging, we can ask them to turn it on, and they will see if one of their render calls is causing an issue.

**Comment by kekekeks** [comment]
>I don't like this kind of validation. We need a way to have these checks only with debug builds and don't magicly set values to 0 and instead throw an exception.

The lack of such checks causes horrible glitches across the entire window.

**Comment by kekekeks** [comment]
Also throwing exceptions during the render pass is an extremely bad design since user code can't do anything meaningful to handle or recover from such exceptions.

**Comment by danwalmsley** [comment]
> Also throwing exceptions during the render pass is an extremely bad design since user code can't do anything meaningful to handle or recover from such exceptions.

Probably wpf gets away with it because direct2d handles invalid calls for them... I checked WPF source and they are not doing any checks inside drawingcontext code...

**Comment by danwalmsley** [comment]
There is a particularly sensitive part of the code, and that is where Image class has 0,0 width and height, it will cause:

https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Image.cs#L76

to generate double.NaN.

thats possibly a bug in the Image code, that we have never noticed... because most of the time it will fail silently.

**Comment by grokys** [comment]
> Also throwing exceptions during the render pass is an extremely bad design since user code can't do anything meaningful to handle or recover from such exceptions.

Is that what's happening? Skia is throwing exceptions?

**Comment by danwalmsley** [comment]
BTW here is an example of a bug this has revealed that has been with us a very long time...
https://github.com/AvaloniaUI/Avalonia/pull/3902

**Comment by danwalmsley** [comment]
> > Also throwing exceptions during the render pass is an extremely bad design since user code can't do anything meaningful to handle or recover from such exceptions.
> 
> Is that what's happening? Skia is throwing exceptions?

if it is I cant catch them... but maybe id need to have compiled skia myself to know...

**Comment by danwalmsley** [comment]
> > > Also throwing exceptions during the render pass is an extremely bad design since user code can't do anything meaningful to handle or recover from such exceptions.
> > 
> > 
> > Is that what's happening? Skia is throwing exceptions?
> 
> if it is I cant catch them... but maybe id need to have compiled skia myself to know... I will have a look.

image

This is my exceptions settings whist the issue is reproducing... nothing, and iv made sure Just My Code is disabled too.

**Comment by danwalmsley** [comment]
@Gillibald if your worried about performance, these checks are happening on the UI thread where all kinds of user code would be doing similar stuff.

its not on the render thread, so it shouldnt cause any measurable performance hit.

**Comment by danwalmsley** [comment]
This is an example of an issue that may or may not be caused by this...

https://youtu.be/C_5la8wpbDA

**Comment by Gillibald** [comment]
Why is `Debug.Assert` a bad thing? Wpf has this all over the code base. This is only used in debug builds.

**Comment by grokys** [comment]
`Debug.Assert` is a very bad thing, because it pops up dialog boxes when running unit tests.

**Comment by danwalmsley** [comment]
tried a repro with just skia, no issues.

tried using immediate rendering and it went away.

updated repro here:
https://github.com/AvaloniaUI/Avalonia/tree/repro-invalid-draw-calls

seems the invalid values affect deferred renderer and layers... 

layers because the repro button flickers, and if you replace it with a simple test control the fillsrectangle with red, no flickering.. possibly the button is in  a different layer.

**Comment by danwalmsley** [comment]
Ok so this is what happens:
1) draw operation takes place with invalid numbers (nan, infinity)
2) this added visual nodes to the scene.
3) deferred renderer then comes to clear nodes here:
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Visuals/Rendering/DeferredRenderer.cs#L428

if the rect to clear has invalid numbers it just clears random areas of the ui.

**Comment by danwalmsley** [comment]
@grokys I think now that we understand what happens we have 2 options...

we use something like in this PR to prevent invalid data getting to the scenegraph...

or we make deferred renderer more robust....

I believe this is better, because it wont affect render performance, and we get the benefit of allowing warnings to help developers.

**Pull request closed without merging**